node/subsystem/src/util.rs

```diff
@@ -0,0 +1,570 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility module for subsystems
+//!
+//! Many subsystems have common interests such as canceling a bunch of spawned jobs,
+//! or determining what their validator ID is. These common interests are factored into
+//! this module.
+
+use crate::{
+	messages::{AllMessages, RuntimeApiMessage, RuntimeApiRequest, SchedulerRoster},
+	FromOverseer, SpawnedSubsystem, Subsystem, SubsystemContext, SubsystemResult,
+};
+use futures::{
+	channel::{mpsc, oneshot},
+	future::Either,
+	prelude::*,
+	select,
+	task::{Spawn, SpawnError, SpawnExt},
+};
+use futures_timer::Delay;
+use keystore::KeyStorePtr;
+use parity_scale_codec::Encode;
+use polkadot_primitives::v1::{
+	EncodeAs, Hash, HeadData, Id as ParaId, Signed, SigningContext,
+	ValidatorId, ValidatorIndex, ValidatorPair,
+};
+use sp_core::Pair;
+use std::{
+	collections::HashMap,
+	convert::{TryFrom, TryInto},
+	pin::Pin,
+	time::Duration,
+};
+use streamunordered::{StreamUnordered, StreamYield};
+
+/// Duration a job will wait after sending a stop signal before hard-aborting.
+pub const JOB_GRACEFUL_STOP_DURATION: Duration = Duration::from_secs(1);
+/// Capacity of channels to and from individual jobs
+pub const JOB_CHANNEL_CAPACITY: usize = 64;
+
+#[derive(Debug, derive_more::From)]
+pub enum Error {
```

Comment:
all public items need doc-strings - this includes enum variants and struct fields.

Reply:
11772f58